File: langs/nodejs/forgeh/src/params.js
// ForgeHash-B3 v1 — parameter definitions and validation (SPECIFICATION.md §7).

export const DEFAULT_OUTPUT_LENGTH = 32;
export const ABSOLUTE_MAX_PARALLELISM = 255;
export const MIN_MEMORY_KIB = 8192;
export const MAX_MEMORY_KIB = 1_048_576;
export const MIN_ITERATIONS = 1;
export const MAX_ITERATIONS = 20;
export const MAX_PARALLELISM_POLICY = 64;
export const MIN_SALT = 16;
export const MAX_SALT = 64;
export const MIN_OUTPUT = 16;
export const MAX_OUTPUT = 64;
export const MAX_PASSWORD = 1_048_576;

/**
 * ForgeHash-B3 parameters. Mirrors the reference `ForgeHashParameters` /
 * Rust `Params` struct.
 */
export class Params {
  constructor({
    memoryKiB = 65_536,
    iterations = 3,
    parallelism = 1,
    outputLength = DEFAULT_OUTPUT_LENGTH,
    saltLength = 16,
  } = {}) {
    this.memoryKiB = memoryKiB;
    this.iterations = iterations;
    this.parallelism = parallelism;
    this.outputLength = outputLength;
    this.saltLength = saltLength;
  }

  /** Recommended interactive profile: 64 MiB, 3 iterations, 1 lane. */
  static interactive() {
    return new Params({
      memoryKiB: 65_536,
      iterations: 3,
      parallelism: 1,
      outputLength: DEFAULT_OUTPUT_LENGTH,
      saltLength: 16,
    });
  }

  /** Development-only profile: fast, low cost. Never use in production. */
  static development() {
    return new Params({
      memoryKiB: MIN_MEMORY_KIB,
      iterations: 1,
      parallelism: 1,
      outputLength: DEFAULT_OUTPUT_LENGTH,
      saltLength: 16,
    });
  }

  /** Recommended sensitive profile: 256 MiB, 4 iterations, 2 lanes. */
  static sensitive() {
    return new Params({
      memoryKiB: 262_144,
      iterations: 4,
      parallelism: 2,
      outputLength: DEFAULT_OUTPUT_LENGTH,
      saltLength: 16,
    });
  }

  /** Throws a descriptive Error if any parameter is out of range. */
  validate() {
    if (
      !Number.isInteger(this.memoryKiB) ||
      this.memoryKiB < MIN_MEMORY_KIB ||
      this.memoryKiB > MAX_MEMORY_KIB
    ) {
      throw new RangeError("ForgeHash: memory out of range");
    }
    if (
      !Number.isInteger(this.iterations) ||
      this.iterations < MIN_ITERATIONS ||
      this.iterations > MAX_ITERATIONS
    ) {
      throw new RangeError("ForgeHash: iterations out of range");
    }
    if (
      !Number.isInteger(this.parallelism) ||
      this.parallelism < 1 ||
      this.parallelism > MAX_PARALLELISM_POLICY ||
      this.parallelism > ABSOLUTE_MAX_PARALLELISM
    ) {
      throw new RangeError("ForgeHash: parallelism out of range");
    }
    if (this.memoryKiB < this.parallelism * 8) {
      throw new RangeError("ForgeHash: memory too small for lane count");
    }
    if (this.memoryKiB % this.parallelism !== 0) {
      throw new RangeError("ForgeHash: memory not divisible by parallelism");
    }
    const blocksPerLane = this.memoryKiB / this.parallelism;
    if (blocksPerLane % 4 !== 0) {
      throw new RangeError("ForgeHash: blocks per lane not divisible by 4");
    }
    if (
      !Number.isInteger(this.outputLength) ||
      this.outputLength < MIN_OUTPUT ||
      this.outputLength > MAX_OUTPUT
    ) {
      throw new RangeError("ForgeHash: output length out of range");
    }
    if (
      !Number.isInteger(this.saltLength) ||
      this.saltLength < MIN_SALT ||
      this.saltLength > MAX_SALT
    ) {
      throw new RangeError("ForgeHash: salt length out of range");
    }
  }
}

File: langs/rust/forgeh/src/params.rs
use crate::error::Error;

pub const DEFAULT_OUTPUT_LENGTH: usize = 32;
pub const ABSOLUTE_MAX_PARALLELISM: usize = 255;
pub const MIN_MEMORY_KIB: usize = 8192;
pub const MAX_MEMORY_KIB: usize = 1_048_576;
pub const MIN_ITERATIONS: usize = 1;
pub const MAX_ITERATIONS: usize = 20;
pub const MAX_PARALLELISM_POLICY: usize = 64;
pub const MIN_SALT: usize = 16;
pub const MAX_SALT: usize = 64;
pub const MIN_OUTPUT: usize = 16;
pub const MAX_OUTPUT: usize = 64;
pub const MAX_PASSWORD: usize = 1_048_576;

#[derive(Clone, Debug)]
pub struct Params {
    pub memory_kib: usize,
    pub iterations: usize,
    pub parallelism: usize,
    pub output_length: usize,
    pub salt_length: usize,
}

impl Default for Params {
    fn default() -> Self {
        Self::interactive()
    }
}

impl Params {
    pub fn interactive() -> Self {
        Self {
            memory_kib: 65_536,
            iterations: 3,
            parallelism: 1,
            output_length: DEFAULT_OUTPUT_LENGTH,
            salt_length: 16,
        }
    }

    pub fn development() -> Self {
        Self {
            memory_kib: MIN_MEMORY_KIB,
            iterations: 1,
            parallelism: 1,
            output_length: DEFAULT_OUTPUT_LENGTH,
            salt_length: 16,
        }
    }

    pub fn sensitive() -> Self {
        Self {
            memory_kib: 262_144,
            iterations: 4,
            parallelism: 2,
            output_length: DEFAULT_OUTPUT_LENGTH,
            salt_length: 16,
        }
    }

    pub fn validate(&self) -> Result<(), Error> {
        if self.memory_kib < MIN_MEMORY_KIB || self.memory_kib > MAX_MEMORY_KIB {
            return Err(Error::Params("memory out of range"));
        }
        if self.iterations < MIN_ITERATIONS || self.iterations > MAX_ITERATIONS {
            return Err(Error::Params("iterations out of range"));
        }
        if self.parallelism < 1
            || self.parallelism > MAX_PARALLELISM_POLICY
            || self.parallelism > ABSOLUTE_MAX_PARALLELISM
        {
            return Err(Error::Params("parallelism out of range"));
        }
        if self.memory_kib < self.parallelism * 8 {
            return Err(Error::Params("memory too small for lanes"));
        }
        if self.memory_kib % self.parallelism != 0 {
            return Err(Error::Params("memory not divisible by parallelism"));
        }
        let blocks_per_lane = self.memory_kib / self.parallelism;
        if blocks_per_lane % 4 != 0 {
            return Err(Error::Params("blocks per lane not divisible by 4"));
        }
        if self.output_length < MIN_OUTPUT || self.output_length > MAX_OUTPUT {
            return Err(Error::Params("output length out of range"));
        }
        if self.salt_length < MIN_SALT || self.salt_length > MAX_SALT {
            return Err(Error::Params("salt length out of range"));
        }
        Ok(())
    }
}
